### ISSUE TITLE
dTPM is not being detected/recognized in OS

### ISSUE THREAD

**Issue by MykeHalk** [Dasharo/dasharo-issues#1083]
### Component

Dasharo firmware

### Device

MSI Pro Z790-P

### Dasharo version

v0.9.2-rc0 / v0.9.2-rc1 latest with 0x12B microcode

### Brief summary

Installed latest tag v0.9.2-rc0 with 0x12B microcode, as well as now removed v0.9.2-rc1 tag, dTPM is recognized in firmware, but not recognized in OS

### How reproducible

100%

### How to reproduce

Install dTPM onto JTPM1 header. Disable ME/ set HAP bit.
Attempt to Install Windows 11 on firmware v0.9.2-rc0/rc1, It wont allow due to TPM requirement.
You can bypass requirements using Rufus bootable drive creator or install using fTPM.
But in OS, Windows "Security Processor" window does not detect TPM. HWinfo also does not detect any TPM.

### Expected behavior

dTPM should be recognized when booting OS

### Actual behavior

dTPM is not recognized

### Screenshots

![Screenshot 2024-10-10 035148](https://github.com/user-attachments/assets/7fbcbc4f-a1f2-4eb8-a233-9f6e7ab33fd3)
![Screenshot 2024-10-10 035212](https://github.com/user-attachments/assets/f541b0af-584a-4f0e-9e8f-0284b604f558)
![Screenshot 2024-10-10 035303](https://github.com/user-attachments/assets/37bce041-ac12-448a-99a7-cef32de7808e)

### Additional context

EDIT: Attempted on Windows 10 as well with same results.

Previous known working branch/tag is `dasharo-4.21`

**Comment by SergiiDmytruk** [comment]
Don't know why `msi_ms7e06_v0.9.2-rc1` was removed, might have happened by accident.  There is no real difference between RC0 and RC1 (RC0 is needed for tests to have something to update to RC1), so don't bother checking both.

Obvious reason could be the change for dTPM, but fTPM is recognized by Windows/Linux when ME is enabled in my tests performed just now (can't test with dTPM).  So this looks weird.

**Comment by zirblazer** [comment]
> @SergiiDmytruk, so it seems that after rebuilding and flashing the firmware, fTPM does work. but as soon as you enable the HAP bit just once, all TPM becomes unusable.
> 
> Once HAP is enabled once, even if you disable it again, fTPM no longer works. It seems this issue is causing dTPM to not work as well.

fTPM requires ME. If you enable HAP, it doesn't work any longer. That is working as intended, and it should also work the same way on previous versions.

dTPM should work with ME disabled (Actually on previous versions with HAP enabled it defaulted to use dTPM if available). Do you have a dTPM installed?

**Comment by pietrushnic** [comment]
> Downgrading to v0.9.1 from branch dasharo-4.21 fixes the issue but I feel that this should be addressed for upcoming release.

This seems to ask for bisection because it sounds like regression.

**Comment by SergiiDmytruk** [comment]
> as soon as you enable the HAP bit just once, all TPM becomes unusable.

When I tested TPM initially wasn't present because ME was disabled, so I had to enable it.  Repeated the test now with toggling ME and TPM is still detected.  Probably needs more testing, but can you also say how you flash?  Random stuff like this reminds me of a confused ME.  Have you tried unplugging your PC from the wall for couple minutes to make sure it ME does a boot cycle?

**Comment by SergiiDmytruk** [comment]
Forgot to mention here that this got dropped in rc2 (yes, in part because I messed it up in rc1).

**Comment by SergiiDmytruk** [comment]
> @SergiiDmytruk I'm confused, will this still be fixed/added in the next release candidate or the main release?

Any follow-up RCs or v1.1.4 & v0.9.2 releases won't have dTPM support.

**Comment by miczyg1** [comment]
The whole TPM unification was meant for these particular MSI boards. It took us weeks/months to develop and upstream it. I can not say I am happy with this turn of events.

**Comment by macpijan** [comment]
v1.1.4-rc3 results with ME HAP-disabled:

![image](https://github.com/user-attachments/assets/a3b18af2-4854-499d-9ee5-4f74cbdc06a9)

![image](https://github.com/user-attachments/assets/1c4ce8d1-d195-4fbd-b0d3-b043e2242c3a)

Is it as expected @miczyg1  ?

**Comment by miczyg1** [comment]
> v1.1.4-rc3 results with ME HAP-disabled:
> 
> ![image](https://private-user-images.githubusercontent.com/19353650/383744567-a3b18af2-4854-499d-9ee5-4f74cbdc06a9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA5NzM4MjIsIm5iZiI6MTczMDk3MzUyMiwicGF0aCI6Ii8xOTM1MzY1MC8zODM3NDQ1NjctYTNiMThhZjItNDg1NC00OTlkLTllZTUtNGY3NGNiZGMwNmE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA3VDA5NTg0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNhYmI5NWQ0N2QwNDdmOThkMDEyMjZlZjY0ZmVmZGE4Mzk2MjdkODYzZjA1ODg2ZGQ3NjNiOGJmMWU5ZGU0ZjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.lvqMnb-adI4p-8pjcpWRZvYCPXTVcr1uIBs5ZESZvQc)
> 
> ![image](https://private-user-images.githubusercontent.com/19353650/383744815-1c4ce8d1-d195-4fbd-b0d3-b043e2242c3a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA5NzM4MjIsIm5iZiI6MTczMDk3MzUyMiwicGF0aCI6Ii8xOTM1MzY1MC8zODM3NDQ4MTUtMWM0Y2U4ZDEtZDE5NS00ZmJkLWIwZDMtYjA0M2UyMjQyYzNhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA3VDA5NTg0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY1OWY3YjQ2ZGVhZjI1MjI0M2Y4MjYxNGYzYzljOTg0NTI3YWRmNWRlYTk3YjFkMjAyMWIxOTMxYmI5ZDllODMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.f-b7xCO1g9X1q6VebL1kg2H35wIunq_wwSMlBPKgnnQ)
> 
> with ME enabled:
> 
> ![image](https://private-user-images.githubusercontent.com/19353650/383745316-8fa797d6-8baa-46ee-ba21-bbb4a023b2d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA5NzM4MjIsIm5iZiI6MTczMDk3MzUyMiwicGF0aCI6Ii8xOTM1MzY1MC8zODM3NDUzMTYtOGZhNzk3ZDYtOGJhYS00NmVlLWJhMjEtYmJiNGEwMjNiMmQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA3VDA5NTg0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU4YzJkYWMxYWZkMGM5NDhjMDg3NDdmZmQ5MjZkNmY4MTY5YTEwNWM2ODRlYjdhNTIzZDE1ZGFmNTk4ZDY1YTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.IgFnqfNzmIjumbR8nEnPf1SBdI7AU0VDF4dgnAx9XBA)
> 
> Is it as expected @miczyg1 ?

Yes, take a look at HID, when ME is disabled it is IFX (Infineon), while ME enabled it is INTC (Intel Corporation)
